### ISSUE TITLE
typeclass search that succeeds then fails after addition of local superclass instance

### ISSUE THREAD

**Issue by jonleivent** [coq/coq#13753]
In Coq 8.12.2

A typeclass search that finds a global instance can then be blocked from finding any instance by the addition of a local superclass instance to the goal context:

```Coq
Class Foo := {}.

Instance foo1 : Foo := {}.

Class Bar`{Foo} := {}.

Instance bar1 : Bar := {}.

Goal True.
  let x:=constr:(_:Bar) in idtac x. (*prints (bar1 : Bar)*)
  assert Foo as foo2 by constructor.
  Fail let x:=constr:(_:Bar) in idtac x.
```
The failure is:
```
The command has indeed failed with message:
Unable to satisfy the following constraints:
In environment:
foo2 : Foo

?H : "Foo"

?b : "Bar"
```

**Comment by jonleivent** [comment]
The bug does not seem to occur when the superclass has informative fields.  For instance, add the field `u:unit` to `Foo` (and correspondingly filling in that field for foo instances), and the bug disappears.  But, if the field `t:True` is used instead, the bug remains.

**Comment by JasonGross** [comment]
> The bug does not seem to occur when the superclass has informative fields.

The fields don't matter, but the sort of the class does; changing it to `Class Foo : Set := {}.` is enough to get the command to pass.  As for why this occurs, we may have to summon @mattam82 .  I suspect that typeclass resolution assumes that the fill for any `Prop`-sorted class cannot matter to the ability to fill other classes and hence disables backtracking into `Prop`-sorted class resolutions.

**Comment by jonleivent** [comment]
@JasonGross How can this be a backtracking issue?  I don't see where backtracking is required.  Also, which class is `Prop`-sorted?

Is typeclass search trying to create a new `Bar` instance from the local `foo2`?  Firstly, it should be able to do so easily - but secondly, it shouldn't be trying to, as `bar1` already exists and should satisfy `(_:Bar)` without any search for `Foo`s .

Note that if the second superclass (`Foo`) instance is global instead of in the goal context, the problem does not occur.

**Comment by SkySkimmer** [comment]
> I suspect that typeclass resolution assumes that the fill for any Prop-sorted class cannot matter to the ability to fill other classes and hence disables backtracking into Prop-sorted class resolutions.

Yup, instances for a Prop typeclass are treated as though Typeclasses Unique Instances was enabled when declaring them
https://github.com/coq/coq/blob/58a4f645ee6d306cb824c2ac2dfa21e460b9692a/tactics/class_tactics.ml#L563-L570
since 1011266b84371b34536dd5aa5afb3a44b8f8d53c AFAICT, although the commit message doesn't mention it.

**Comment by jonleivent** [comment]
So `Class Foo:Type := {}` is unique but `Class Foo:Set := {}` is not?

**Comment by JasonGross** [comment]
Yes, but only because Coq is strange about minimizing `Type` to `Prop`:
```coq
Class Foo : Type := {}.
Check Foo. (* Foo : Prop *)
```

**Comment by JasonGross** [comment]
If instead you do
```coq
Universe i.
Class Foo : Type@{i} := {}.
```
then you get a non-unique class.

**Comment by jonleivent** [comment]
> Yes, but only because Coq is strange about minimizing Type to Prop
Oh, that sounds familiar.  I probably knew that once, and it got partially GC'ed.

**Comment by JasonGross** [comment]
> @JasonGross How can this be a backtracking issue? I don't see where backtracking is required. Also, which class is `Prop`-sorted?

The `Prop`-sorted class is `Foo`.  Consider the typeclass log of `Set Typeclasses Debug. Set Printing All.` when `Foo : Prop`:
```
Debug: 1: looking for Foo without backtracking
Debug: 1.1: exact foo2 on Foo, 0 subgoal(s)
Debug: 2: looking for (@Bar foo2) without backtracking
Debug: 2: no match for (@Bar foo2), 1 possibilities
Debug: 1: looking for Foo without backtracking
Debug: 1.1: exact foo2 on Foo, 0 subgoal(s)
Debug: 2: looking for (@Bar foo2) without backtracking
Debug: 2: no match for (@Bar foo2), 1 possibilities
```
(no idea why things are repeated twice) vs when `Foo : Set`:
```
Debug: 1: looking for Foo with backtracking
Debug: 1.1: exact foo2 on Foo, 0 subgoal(s)
Debug: 2: looking for (@Bar foo2) without backtracking
Debug: 2: no match for (@Bar foo2), 1 possibilities
Debug: 1.2: exact foo1 on Foo, 0 subgoal(s)
Debug: 2: looking for (@Bar foo1) without backtracking
Debug: 2.1: exact bar1 on (@Bar foo1), 0 subgoal(s)
Debug: Finished run 1 of resolution.
```

> Note that if the second superclass (`Foo`) instance is global instead of in the goal context, the problem does not occur.

This is because your global instance is transparent whereas your local one is not.  If you instead do
```coq
Axiom foo2 : Foo.
Existing Instance foo2.
```
instead of using `Instance foo2 : Foo`, the discrepancy should go away (i.e., this will make TC resolution fail).

**Comment by jonleivent** [comment]
@JasonGross that debug log is very interesting.  It suggests the bug is cost dependent.  If the blocking instance of Foo is given a high enough cost, it stops blocking the search from finding bar1.  Yup, that's what happens.

**Comment by jonleivent** [comment]
But I don't know why the opaqueness of `foo2` should matter.  I agree that it does - doing `Typeclasses Opaque foo2` on a global `foo2` also causes the bug.

**Comment by JasonGross** [comment]
If there is a bug here, the bug is that `Prop`-sorted typeclass instances are treated as unique.

The thing that happens is that typeclass resolution tries to first resolve `?1 : Foo` in `@Bar ?1` and then tries to resolve `?2 : @Bar ?1`.  If backtracking does not happen, then `?1` gets resolved to the most-recent least-cost instance, in this case `foo2`.  If `foo2` is convertible with `foo1`, then `bar1` is considered an instance of `@Bar foo2` as well as `@Bar foo1`.  Alternatively, if backtracking is enabled, then we can backtrack and instead choose `foo1` for `?1`.  However, if backtracking is disabled and `foo1` is not convertible with `foo2`, then typeclass resolution fails when it sees that `bar1` is not an instance of `@Bar foo2`.

**Comment by jonleivent** [comment]
Also `pose Build_foo as foo3` in the goal doesn't cause the bug, again because it isn't opaque.   Is this a more general bug that opaqueness interferes with uniqueness considerations?

**Comment by JasonGross** [comment]
> Is this a more general bug that opaqueness interferes with uniqueness considerations?

It's not a bug.  Declaring an instance unique is only accurate if the instance is in fact unique (up to judgmental equality without unfolding things that are declared opaque by the typeclass hint database).  Coq does not check nor validate unique instances in any way, it just trusts you that they are unique.

**Comment by SkySkimmer** [comment]
>Also pose Build_foo as foo3 in the goal doesn't cause the bug, again because it isn't opaque. Is this a more general bug that opaqueness interferes with uniqueness considerations?

If `foo2` is opaque then `bar1` is not an instance of `Bar foo2`, but if it's transparent then it works.
